File: apps/ios/Ash/Ash/Core/Services/RelayService.swift
//
//  RelayService.swift
//  Ash
//

import Foundation

protocol RelayServiceProtocol: Sendable {
    func submitMessage(
        conversationId: String,
        authToken: String,
        ciphertext: Data,
        sequence: UInt64?,
        ttlSeconds: UInt64?,
        extendedTTL: Bool,
        persistent: Bool
    ) async throws -> UUID

    func pollMessages(
        conversationId: String,
        authToken: String,
        cursor: RelayCursor?
    ) async throws -> PollResponse

    /// Acknowledge message delivery (tells sender we received it)
    func ackMessages(
        conversationId: String,
        authToken: String,
        blobIds: [UUID]
    ) async throws -> Int

    func registerDevice(
        conversationId: String,
        authToken: String,
        deviceToken: String
    ) async throws

    func registerConversation(
        conversationId: String,
        authTokenHash: String,
        burnTokenHash: String
    ) async throws

    func burnConversation(conversationId: String, burnToken: String) async throws

    func checkBurnStatus(conversationId: String, authToken: String) async throws -> BurnStatus
}

typealias RelayCursor = String

struct PollResponse: Sendable {
    let messages: [RelayMessage]
    let nextCursor: RelayCursor?
    let burned: Bool
}

struct RelayMessage: Sendable {
    let id: UUID
    let sequence: UInt64?
    let ciphertext: Data
    let receivedAt: Date
}

struct BurnStatus: Sendable {
    let burned: Bool
    let burnedAt: Date?
}

enum RelayError: Error, Sendable {
    case invalidURL
    case networkError(Error)
    case serverError(statusCode: Int, message: String?)
    case decodingError(Error)
    case conversationBurned
    case conversationNotFound
    case payloadTooLarge
    case queueFull
    case noConnection
}

extension RelayError: CustomStringConvertible {
    var description: String {
        switch self {
        case .invalidURL: return "Invalid URL"
        case .networkError(let error): return "Network: \(error.localizedDescription)"
        case .serverError(let code, let message): return "Server \(code): \(message ?? "unknown")"
        case .decodingError(let error): return "Decode: \(error)"
        case .conversationBurned: return "Conversation burned"
        case .conversationNotFound: return "Conversation not registered"
        case .payloadTooLarge: return "Payload too large"
        case .queueFull: return "Queue full"
        case .noConnection: return "No connection"
        }
    }
}

final class RelayService: RelayServiceProtocol, Sendable {
    private let baseURL: URL
    private let session: URLSession
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    private func logId(_ conversationId: String) -> String {
        String(conversationId.prefix(8))
    }

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session

        self.encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601

        self.decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let dateString = try container.decode(String.self)

            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = formatter.date(from: dateString) {
                return date
            }

            formatter.formatOptions = [.withInternetDateTime]
            if let date = formatter.date(from: dateString) {
                return date
            }

            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Cannot decode date: \(dateString)"
            )
        }

        Log.debug(.relay, "Relay service initialized: \(baseURL.host ?? "unknown")")
    }

    convenience init(baseURLString: String) throws {
        guard let url = URL(string: baseURLString) else {
            throw RelayError.invalidURL
        }
        self.init(baseURL: url)
    }

    func submitMessage(
        conversationId: String,
        authToken: String,
        ciphertext: Data,
        sequence: UInt64?,
        ttlSeconds: UInt64?,
        extendedTTL: Bool,
        persistent: Bool
    ) async throws -> UUID {
        let url = baseURL.appendingPathComponent("v1/messages")
        let id = logId(conversationId)

        Log.debug(.relay, "[\(id)] Submitting: \(ciphertext.count) bytes, seq=\(sequence ?? 0), ttl=\(ttlSeconds ?? 0)s")

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let body = SubmitMessageRequest(
            conversationId: conversationId,
            ciphertext: ciphertext.base64EncodedString(),
            sequence: sequence,
            ttlSeconds: ttlSeconds,
            extendedTTL: extendedTTL,
            persistent: persistent
        )
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await performRequest(request)
        try validateResponse(response, data: data)

        let result = try decoder.decode(SubmitMessageResponse.self, from: data)

        guard result.accepted else {
            Log.error(.relay, "[\(id)] Message rejected by server")
            throw RelayError.serverError(statusCode: 400, message: "Message not accepted")
        }

        Log.debug(.relay, "[\(id)] Submitted successfully: blob=\(result.blobId.uuidString.prefix(8))")
        return result.blobId
    }

    func pollMessages(
        conversationId: String,
        authToken: String,
        cursor: RelayCursor?
    ) async throws -> PollResponse {
        var components = URLComponents(url: baseURL.appendingPathComponent("v1/messages"), resolvingAgainstBaseURL: true)!
        var queryItems = [URLQueryItem(name: "conversation_id", value: conversationId)]
        if let cursor = cursor {
            queryItems.append(URLQueryItem(name: "cursor", value: cursor))
        }
        components.queryItems = queryItems

        let id = logId(conversationId)

        var request = URLRequest(url: components.url!)
        request.httpMethod = "GET"
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await performRequest(request)
        try validateResponse(response, data: data)

        let result: PollMessagesResponse
        do {
            result = try decoder.decode(PollMessagesResponse.self, from: data)
        } catch {
            Log.error(.relay, "[\(id)] Failed to decode poll response: \(error)")
            throw RelayError.decodingError(error)
        }

        if !result.messages.isEmpty {
            Log.debug(.relay, "[\(id)] Poll returned \(result.messages.count) messages, burned=\(result.burned)")
        }

        let messages = result.messages.compactMap { msg -> RelayMessage? in
            guard let ciphertext = Data(base64Encoded: msg.ciphertext) else {
                Log.error(.relay, "[\(id)] Invalid base64 in message \(msg.id.uuidString.prefix(8))")
                return nil
            }
            return RelayMessage(
                id: msg.id,
                sequence: msg.sequence,
                ciphertext: ciphertext,
                receivedAt: msg.receivedAt
            )
        }

        return PollResponse(
            messages: messages,
            nextCursor: result.nextCursor,
            burned: result.burned
        )
    }

    func ackMessages(
        conversationId: String,
        authToken: String,
        blobIds: [UUID]
    ) async throws -> Int {
        guard !blobIds.isEmpty else { return 0 }

        let url = baseURL.appendingPathComponent("v1/messages/ack")
        let id = logId(conversationId)

        Log.debug(.relay, "[\(id)] Acknowledging \(blobIds.count) messages")

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let body = AckMessageRequest(
            conversationId: conversationId,
            blobIds: blobIds
        )
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await performRequest(request)
        try validateResponse(response, data: data)

        let result = try decoder.decode(AckMessageResponse.self, from: data)
        Log.debug(.relay, "[\(id)] Acknowledged \(result.acknowledged) messages")
        return result.acknowledged
    }

    func registerDevice(
        conversationId: String,
        authToken: String,
        deviceToken: String
    ) async throws {
        let url = baseURL.appendingPathComponent("v1/register")
        let id = logId(conversationId)

        Log.info(.relay, "[\(id)] Registering device for push notifications")

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let body = RegisterDeviceRequest(
            conversationId: conversationId,
            deviceToken: deviceToken,
            platform: "ios"
        )
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await performRequest(request)
        try validateResponse(response, data: data)

        Log.info(.relay, "[\(id)] Device registered successfully")
    }

    func registerConversation(
        conversationId: String,
        authTokenHash: String,
        burnTokenHash: String
    ) async throws {
        let url = baseURL.appendingPathComponent("v1/conversations")
        let id = logId(conversationId)

        Log.info(.relay, "[\(id)] Registering conversation with relay")

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = RegisterConversationRequest(
            conversationId: conversationId,
            authTokenHash: authTokenHash,
            burnTokenHash: burnTokenHash
        )
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await performRequest(request)
        try validateResponse(response, data: data)

        Log.info(.relay, "[\(id)] Conversation registered successfully")
    }

    func burnConversation(conversationId: String, burnToken: String) async throws {
        let url = baseURL.appendingPathComponent("v1/burn")
        let id = logId(conversationId)

        Log.warning(.relay, "[\(id)] Burning conversation on relay")

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = BurnConversationRequest(conversationId: conversationId, burnToken: burnToken)
        request.httpBody = try encoder.encode(body)

        let (data, response) = try await performRequest(request)
        try validateResponse(response, data: data)

        Log.warning(.relay, "[\(id)] Conversation burned on relay")
    }

    func checkBurnStatus(conversationId: String, authToken: String) async throws -> BurnStatus {
        var components = URLComponents(url: baseURL.appendingPathComponent("v1/burn"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "conversation_id", value: conversationId)]

        var request = URLRequest(url: components.url!)
        request.httpMethod = "GET"
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await performRequest(request)
        try validateResponse(response, data: data)

        let result = try decoder.decode(BurnStatusResponse.self, from: data)

        if result.burned {
            Log.warning(.relay, "[\(logId(conversationId))] Conversation marked as burned")
        }

        return BurnStatus(
            burned: result.burned,
            burnedAt: result.burnedAt
        )
    }

    private func performRequest(_ request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: request)
        } catch {
            Log.error(.relay, "Network request failed: \(error.localizedDescription)")
            throw RelayError.networkError(error)
        }
    }

    private func validateResponse(_ response: URLResponse, data: Data) throws {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw RelayError.networkError(URLError(.badServerResponse))
        }

        switch httpResponse.statusCode {
        case 200...299:
            return
        case 400:
            if let errorResponse = try? decoder.decode(ErrorResponse.self, from: data) {
                if errorResponse.error.contains("burned") {
                    throw RelayError.conversationBurned
                } else if errorResponse.error.contains("too large") {
                    throw RelayError.payloadTooLarge
                } else if errorResponse.error.contains("full") {
                    throw RelayError.queueFull
                }
                throw RelayError.serverError(statusCode: 400, message: errorResponse.error)
            }
            throw RelayError.serverError(statusCode: 400, message: nil)
        case 404:
            // Conversation not registered with relay - needs re-registration
            if let errorResponse = try? decoder.decode(ErrorResponse.self, from: data),
               errorResponse.error.lowercased().contains("not found") {
                throw RelayError.conversationNotFound
            }
            throw RelayError.conversationNotFound
        default:
            let message = String(data: data, encoding: .utf8)
            Log.error(.relay, "Server error \(httpResponse.statusCode): \(message ?? "unknown")")
            throw RelayError.serverError(statusCode: httpResponse.statusCode, message: message)
        }
    }
}

private struct SubmitMessageRequest: Encodable {
    let conversationId: String
    let ciphertext: String
    let sequence: UInt64?
    let ttlSeconds: UInt64?
    let extendedTTL: Bool
    let persistent: Bool

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case ciphertext, sequence
        case ttlSeconds = "ttl_seconds"
        case extendedTTL = "extended_ttl"
        case persistent
    }
}

private struct SubmitMessageResponse: Decodable {
    let accepted: Bool
    let blobId: UUID

    enum CodingKeys: String, CodingKey {
        case accepted
        case blobId = "blob_id"
    }
}

private struct PollMessagesResponse: Decodable {
    let messages: [MessageBlob]
    let nextCursor: RelayCursor?
    let burned: Bool

    enum CodingKeys: String, CodingKey {
        case messages
        case nextCursor = "next_cursor"
        case burned
    }
}

private struct MessageBlob: Decodable {
    let id: UUID
    let sequence: UInt64?
    let ciphertext: String
    let receivedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, sequence, ciphertext
        case receivedAt = "received_at"
    }
}

private struct RegisterDeviceRequest: Encodable {
    let conversationId: String
    let deviceToken: String
    let platform: String

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case deviceToken = "device_token"
        case platform
    }
}

private struct RegisterConversationRequest: Encodable {
    let conversationId: String
    let authTokenHash: String
    let burnTokenHash: String

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case authTokenHash = "auth_token_hash"
        case burnTokenHash = "burn_token_hash"
    }
}

private struct BurnConversationRequest: Encodable {
    let conversationId: String
    let burnToken: String

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case burnToken = "burn_token"
    }
}

private struct BurnStatusResponse: Decodable {
    let burned: Bool
    let burnedAt: Date?

    enum CodingKeys: String, CodingKey {
        case burned
        case burnedAt = "burned_at"
    }
}

private struct ErrorResponse: Decodable {
    let error: String
}

private struct AckMessageRequest: Encodable {
    let conversationId: String
    let blobIds: [UUID]

    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case blobIds = "blob_ids"
    }
}

private struct AckMessageResponse: Decodable {
    let acknowledged: Int
}

File: apps/android/app/src/main/java/com/monadial/ash/core/services/RelayService.kt
package com.monadial.ash.core.services

import android.util.Base64
import android.util.Log
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// === Request DTOs (matching iOS) ===

@Serializable
data class SubmitMessageRequest(
    @SerialName("conversation_id") val conversationId: String,
    val ciphertext: String,
    val sequence: Long? = null,
    @SerialName("ttl_seconds") val ttlSeconds: Long? = null,
    @SerialName("extended_ttl") val extendedTTL: Boolean = false,
    val persistent: Boolean = false
)

@Serializable
data class SubmitMessageResponse(val accepted: Boolean, @SerialName("blob_id") val blobId: String)

@Serializable
data class AckMessageRequest(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("blob_ids") val blobIds: List<String>
)

@Serializable
data class AckMessageResponse(val acknowledged: Int)

@Serializable
data class RegisterConversationRequest(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("auth_token_hash") val authTokenHash: String,
    @SerialName("burn_token_hash") val burnTokenHash: String
)

@Serializable
data class RegisterDeviceRequest(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("device_token") val deviceToken: String,
    val platform: String = "android"
)

@Serializable
data class BurnConversationRequest(
    @SerialName("conversation_id") val conversationId: String,
    @SerialName("burn_token") val burnToken: String
)

@Serializable
data class BurnStatusResponse(val burned: Boolean, @SerialName("burned_at") val burnedAt: String? = null)

// === Response DTOs ===

@Serializable
data class HealthResponse(val status: String, val version: String? = null)

@Serializable
data class PollMessageItem(
    val id: String,
    val ciphertext: String,
    val sequence: Long? = null,
    @SerialName("received_at") val receivedAt: String
)

@Serializable
data class PollMessagesResponse(
    val messages: List<PollMessageItem>,
    @SerialName("next_cursor") val nextCursor: String? = null,
    val burned: Boolean = false
)

// === Result types ===

data class ConnectionTestResult(
    val success: Boolean,
    val version: String? = null,
    val latencyMs: Long? = null,
    val error: String? = null
)

data class SendResult(val success: Boolean, val blobId: String? = null, val error: String? = null)

data class PollResult(
    val success: Boolean,
    val messages: List<ReceivedMessage> = emptyList(),
    val cursor: String? = null,
    val burned: Boolean = false,
    val error: String? = null
)

data class ReceivedMessage(val id: String, val ciphertext: ByteArray, val sequence: Long?, val receivedAt: String) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as ReceivedMessage
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()
}

@Singleton
class RelayService @Inject constructor(
    private val httpClient: HttpClient,
    private val settingsService: SettingsService
) {
    companion object {
        private const val TAG = "RelayService"
    }

    private fun logId(conversationId: String) = conversationId.take(8)

    // === Health Check ===

    suspend fun checkHealth(relayUrl: String? = null): Result<HealthResponse> {
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()
            val response: HttpResponse = httpClient.get("$url/health")
            if (response.status.isSuccess()) {
                Result.success(response.body())
            } else {
                Result.failure(Exception("Health check failed: ${response.status}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun testConnection(relayUrl: String): ConnectionTestResult {
        return try {
            val startTime = System.currentTimeMillis()
            val response: HttpResponse = httpClient.get("$relayUrl/health")
            val latencyMs = System.currentTimeMillis() - startTime
            if (response.status.isSuccess()) {
                val health: HealthResponse = response.body()
                ConnectionTestResult(success = true, version = health.version, latencyMs = latencyMs)
            } else {
                ConnectionTestResult(success = false, error = "Status: ${response.status}")
            }
        } catch (e: Exception) {
            ConnectionTestResult(success = false, error = e.message)
        }
    }

    // === Messages (matching iOS: POST /v1/messages) ===

    suspend fun submitMessage(
        conversationId: String,
        authToken: String,
        ciphertext: ByteArray,
        sequence: Long? = null,
        ttlSeconds: Long? = null,
        extendedTTL: Boolean = false,
        persistent: Boolean = false,
        relayUrl: String? = null
    ): SendResult {
        val id = logId(conversationId)
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()
            val encoded = Base64.encodeToString(ciphertext, Base64.NO_WRAP)

            Log.d(TAG, "[$id] Submitting: ${ciphertext.size} bytes, seq=${sequence ?: 0}, ttl=${ttlSeconds ?: 0}s")

            val request =
                SubmitMessageRequest(
                    conversationId = conversationId,
                    ciphertext = encoded,
                    sequence = sequence,
                    ttlSeconds = ttlSeconds,
                    extendedTTL = extendedTTL,
                    persistent = persistent
                )

            val response: HttpResponse =
                httpClient.post("$url/v1/messages") {
                    header("Authorization", "Bearer $authToken")
                    contentType(ContentType.Application.Json)
                    setBody(request)
                }

            if (response.status.isSuccess()) {
                val result: SubmitMessageResponse = response.body()
                if (result.accepted) {
                    Log.d(TAG, "[$id] Submitted successfully: blob=${result.blobId.take(8)}")
                    SendResult(success = true, blobId = result.blobId)
                } else {
                    SendResult(success = false, error = "Message not accepted")
                }
            } else {
                SendResult(success = false, error = "Status: ${response.status}")
            }
        } catch (e: Exception) {
            Log.e(TAG, "[$id] Submit failed: ${e.message}")
            SendResult(success = false, error = e.message)
        }
    }

    // === Poll Messages (matching iOS: GET /v1/messages?conversation_id=...) ===

    suspend fun fetchMessages(
        conversationId: String,
        authToken: String,
        cursor: String? = null,
        relayUrl: String? = null
    ): PollResult {
        val id = logId(conversationId)
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()
            val response: HttpResponse =
                httpClient.get("$url/v1/messages") {
                    header("Authorization", "Bearer $authToken")
                    parameter("conversation_id", conversationId)
                    cursor?.let { parameter("cursor", it) }
                }

            if (response.status.isSuccess()) {
                val result: PollMessagesResponse = response.body()
                val receivedMessages =
                    result.messages.map { msg ->
                        ReceivedMessage(
                            id = msg.id,
                            ciphertext = Base64.decode(msg.ciphertext, Base64.DEFAULT),
                            sequence = msg.sequence,
                            receivedAt = msg.receivedAt
                        )
                    }
                if (result.messages.isNotEmpty()) {
                    Log.d(TAG, "[$id] Poll returned ${result.messages.size} messages, burned=${result.burned}")
                }
                PollResult(
                    success = true,
                    messages = receivedMessages,
                    cursor = result.nextCursor,
                    burned = result.burned
                )
            } else {
                PollResult(success = false, error = "Status: ${response.status}")
            }
        } catch (e: Exception) {
            PollResult(success = false, error = e.message)
        }
    }

    fun pollMessages(
        conversationId: String,
        authToken: String,
        cursor: String? = null,
        relayUrl: String? = null
    ): Flow<PollResult> = flow {
        var currentCursor = cursor
        while (true) {
            val result = fetchMessages(conversationId, authToken, currentCursor, relayUrl)
            emit(result)
            if (result.success && result.cursor != null) {
                currentCursor = result.cursor
            }
            delay(5000) // Poll every 5 seconds
        }
    }

    // === Acknowledge Messages (matching iOS: POST /v1/messages/ack) ===

    suspend fun acknowledgeMessages(
        conversationId: String,
        authToken: String,
        blobIds: List<String>,
        relayUrl: String? = null
    ): Result<Int> {
        if (blobIds.isEmpty()) return Result.success(0)

        val id = logId(conversationId)
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()

            Log.d(TAG, "[$id] Acknowledging ${blobIds.size} messages")

            val request =
                AckMessageRequest(
                    conversationId = conversationId,
                    blobIds = blobIds
                )

            val response: HttpResponse =
                httpClient.post("$url/v1/messages/ack") {
                    header("Authorization", "Bearer $authToken")
                    contentType(ContentType.Application.Json)
                    setBody(request)
                }

            if (response.status.isSuccess()) {
                val result: AckMessageResponse = response.body()
                Log.d(TAG, "[$id] Acknowledged ${result.acknowledged} messages")
                Result.success(result.acknowledged)
            } else {
                Result.failure(Exception("ACK failed: ${response.status}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // === Register Conversation (matching iOS: POST /v1/conversations) ===

    suspend fun registerConversation(
        conversationId: String,
        authTokenHash: String,
        burnTokenHash: String,
        relayUrl: String? = null
    ): Result<Unit> {
        val id = logId(conversationId)
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()

            Log.d(TAG, "[$id] Registering conversation with relay")

            val request =
                RegisterConversationRequest(
                    conversationId = conversationId,
                    authTokenHash = authTokenHash,
                    burnTokenHash = burnTokenHash
                )

            val response: HttpResponse =
                httpClient.post("$url/v1/conversations") {
                    contentType(ContentType.Application.Json)
                    setBody(request)
                }

            if (response.status.isSuccess()) {
                Log.d(TAG, "[$id] Conversation registered successfully")
                Result.success(Unit)
            } else {
                Result.failure(Exception("Registration failed: ${response.status}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // === Register Device (matching iOS: POST /v1/register) ===

    suspend fun registerDevice(
        conversationId: String,
        authToken: String,
        deviceToken: String,
        relayUrl: String? = null
    ): Result<Unit> {
        val id = logId(conversationId)
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()

            Log.d(TAG, "[$id] Registering device for push notifications")

            val request =
                RegisterDeviceRequest(
                    conversationId = conversationId,
                    deviceToken = deviceToken,
                    platform = "android"
                )

            val response: HttpResponse =
                httpClient.post("$url/v1/register") {
                    header("Authorization", "Bearer $authToken")
                    contentType(ContentType.Application.Json)
                    setBody(request)
                }

            if (response.status.isSuccess()) {
                Log.d(TAG, "[$id] Device registered successfully")
                Result.success(Unit)
            } else {
                Result.failure(Exception("Device registration failed: ${response.status}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // === Burn Conversation (matching iOS: POST /v1/burn) ===

    suspend fun burnConversation(conversationId: String, burnToken: String, relayUrl: String? = null): Result<Unit> {
        val id = logId(conversationId)
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()

            Log.w(TAG, "[$id] Burning conversation on relay")

            val request =
                BurnConversationRequest(
                    conversationId = conversationId,
                    burnToken = burnToken
                )

            val response: HttpResponse =
                httpClient.post("$url/v1/burn") {
                    contentType(ContentType.Application.Json)
                    setBody(request)
                }

            if (response.status.isSuccess()) {
                Log.w(TAG, "[$id] Conversation burned on relay")
                Result.success(Unit)
            } else {
                Result.failure(Exception("Burn failed: ${response.status}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // === Check Burn Status (matching iOS: GET /v1/burn?conversation_id=...) ===

    suspend fun checkBurnStatus(
        conversationId: String,
        authToken: String,
        relayUrl: String? = null
    ): Result<BurnStatusResponse> {
        return try {
            val url = relayUrl ?: settingsService.relayServerUrl.first()
            val response: HttpResponse =
                httpClient.get("$url/v1/burn") {
                    header("Authorization", "Bearer $authToken")
                    parameter("conversation_id", conversationId)
                }
            if (response.status.isSuccess()) {
                Result.success(response.body())
            } else {
                Result.failure(Exception("Burn status check failed: ${response.status}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // === Utility ===

    fun hashToken(token: String): String {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(token.toByteArray(Charsets.UTF_8))
        return hashBytes.joinToString("") { String.format("%02x", it) }
    }
}
